pm: policy: more cleanups

gmarull opened the pull request:
Principal changes in this PR:

- Add `pm_state_cpu_get_all` API to obtain CPU power states information.
- Simplify residency policy by using this new API
- Policy returns now a reference to the next state instead of a struct copy
- Add new tests that cover the behavior of the policy API

Others:
- Added `cpus` node to native_posix and leon3 (was missing), this simplifies PM code as we can assume `cpus` exists
- Fixed some include issues (sys/time_units.h)

gmarull committed:
pm: state: add pm_state_cpu_get_all

Add a new API to obtain the list of power states available for a given
CPU.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: policy: residency: use pm_state_cpu_get_all

Use the pm_state_cpu_get_all API to obtain the list of available CPU
states. This changes reduces the need for non-interesting code within
the policy codebase.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: policy: return a reference to the next state

Return a constant reference to the next state instead of a copy of
struct pm_state_info. When the next state should be active, just return
NULL. Struct copying should be in general avoided, specially in code
paths executed frequently as is this one.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: policy: residency: remove redundant logging

The pm.c already performs sufficient logging/tracing, so cleanup the
policy file.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: serial: ns16550: use pm_state_cpu_get_all

Use pm_state_cpu_get_all to obtain CPU states instead of using extra ROM
duplicating existing information already holded by the PM subsystem.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
sys: time_units: add missing include

The header can't be fully used in standalone mode: toolchain.h has to be
included first, otherwise the ALWAYS_INLINE attribute is not defined.
Headers that can be directly included and are not self-contained should
be considered a bad practice.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
tests: pm: policy_api: test policy behavior

Add a new test to verify the policy API behavior. Test also checks that
custom policies can be implemented.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
dts: bindings: cpu: add compatible for native_posix

Create a CPU compatible for native_posix builds. It can be useful when
testing.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
boards: posix: native_posix: add cpus and cpu0 nodes

Define the cpus node and the first cpu. The cpus node can be used when
testing, e.g. power states.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: state: require cpus node

The PM state code is in practice useless when no cpus are defined in DT,
so require this node to be defined.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

aescolar reviewed (commented):
No complaint for the native_posix changes, but those are not the main point of the PR. So just +0'ing it:)

gmarull commented:
> No complaint for the native_posix changes, but those are not the main point of the PR. So just +0'ing it:)

Are not the main point but are later used/required.

gmarull committed:
dts: sparc: add cpus node to leon3

The leon3 SoCs were missing definitions of the CPU node. This node is
now required for PM, so that power states can be defined.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

ceolin commented on subsys/pm/pm.c:
```diff
@@ -194,7 +194,16 @@ bool pm_system_suspend(int32_t ticks)
+			z_power_states[id] = active;
+		} else {
+			z_power_states[id] = *info;
+		}
```
The current power state is `PM_STATE_ACTIVE`, you don't need the assignment when it policy returns `NULL`

gmarull replied:
This is keeping current behavior: as of today, the state returned by `pm_policy_next_state` is always copied to `z_power_states[id]`, even when it is `PM_STATE_ACTIVE`. I'm planning to improve `pm.c` after this.

ceolin replied:
what I saying is that when this function is called, `z_power_states[id]` is `PM_STATE_ACTIVE`. So, when the policy returns `NULL` you don't need to this assigment.
The previous assignment is done because it was cheaper the assignment than a branch + assignement. But now, how it is necessary the branch to check if the policy returned `NULL`, you no longer it.

ceolin replied:
that makes sense - https://github.com/zephyrproject-rtos/zephyr/pull/41703

Fixed the initialization, the cpu state should always be `ACTIVE` when `pm_system_suspend` is called and it can be simplified to:

````
if (info != NULL) {
    z_power_states[id] = *info;
}
````

gmarull replied:
adjusted, https://github.com/zephyrproject-rtos/zephyr/pull/41703 will make sure that this is correct.

ceolin commented on subsys/pm/state.c:
```diff
@@ -0,0 +1,64 @@
+	BUILD_ASSERT(							       \
+		DT_PROP_BY_PHANDLE_IDX_OR(node_id, cpu_power_states, i,	       \
+					  min_residency_us, 0U) >=	       \
+		DT_PROP_BY_PHANDLE_IDX_OR(node_id, cpu_power_states, i,	       \
```
I believe the assert here should be `min_residency_us > exit_latency_us`. Does it make any sense having `min_residency_us == exit_latency_us` ?

gmarull replied:
This is keeping current behavior, which btw you suggested here: https://github.com/zephyrproject-rtos/zephyr/pull/41124#discussion_r768333685

ceolin replied:
yep yep yep. Missed the assert, it continues to make sense the `>=` :)

ceolin reviewed (changes requested):
Minor fix and it is good to go IMHO.

carlocaione reviewed (approved):
+1 for the code

gmarull commented:
@ceolin review ping, I'd like to make more changes on top of these.

ceolin commented:
> @ceolin review ping, I'd like to make more changes on top of these.

just replied one of the comments. The other is resolved for me.
